interceptors/payment_authentication.go

```diff
@@ -0,0 +1,81 @@
+package interceptors
+
+import (
+	"fmt"
+	"github.com/companieshouse/chs.go/log"
+	"github.com/companieshouse/payments.api.ch.gov.uk/config"
+	"github.com/companieshouse/payments.api.ch.gov.uk/dao"
+	"github.com/companieshouse/payments.api.ch.gov.uk/helpers"
+	"github.com/companieshouse/payments.api.ch.gov.uk/service"
+	"github.com/gorilla/mux"
+	"net/http"
+)
+
+func PaymentAuthenticationInterceptor(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check for a payment ID in request
+		vars := mux.Vars(r)
+		id := vars["payment_id"]
+		if id == "" {
+			log.ErrorR(r, fmt.Errorf("PaymentAuthenticationInterceptor error: no payment id"))
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		cfg, err := config.Get()
+		if err != nil {
+			log.Error(fmt.Errorf("error retreiving config: %s. Exiting", err), nil)
+			return
+		}
+
+		m := &dao.Mongo{
+			URL: cfg.MongoDBURL,
+		}
+
+		p := service.PaymentService{
+			DAO:    m,
+			Config: *cfg,
+		}
+
+		// Get the payment session form the ID in request
+		paymentSession, httpStatus, err := p.GetPaymentSession(id)
+		if err != nil {
+			log.Error(fmt.Errorf("PaymentAuthenticationInterceptor not found: payment session found"))
+			w.WriteHeader(httpStatus)
+			return
+		}
```

Comment:
once we've got the payment session data from payment service looking it up in mongo we should store it in request context for use later on to avoid multiple mongo reads e.g. on GET /payments/{payment_id} this interceptor runs and store payment session data in request context then the handler just returns the payment session data already cached in the request context rather than calling payment service again

Reply:
updated